src/store.cc

```diff
@@ -540,7 +540,17 @@ storeGetPublic(const char *uri, const HttpRequestMethod& method)
 StoreEntry *
 storeGetPublicByRequestMethod(HttpRequest * req, const HttpRequestMethod& method, const KeyScope keyScope)
 {
-    return Store::Root().find(storeKeyPublicByRequestMethod(req, method, keyScope));
+    StoreEntry *e = Store::Root().find(storeKeyPublicByRequestMethod(req, method, keyScope));
```

Comment:
Hiding public cache entries at such a low level is dangerous -- the caller might not be subject to collapsing risks and should get access to the public entry even if that entry has not seen its headers yet. For example, findPreviouslyCachedEntry() in http.cc may need to see the entry regardless of the collapsed_forwarding_access. Also, applying collapsed_forwarding_access logic to some but not all Controller::find() callers may create a distorted/inconsistent view of the cache when a different parts of the same transaction are calling different find() wrappers and getting conflicting "yes, it is here" and "no, it is not cached" results.

I believe `collapsed_forwarding_access deny` means that the matching request must be shielded from CF risks; it must not be enslaved. This goal is a little different from the "must not see the matching initiator transaction" rule currently implemented by this PR code.

If you agree, then please analyze storeGetPublic*() (and related functions) callers and protect those callers that should not be subject to CF risks. For example, I suspect that clientReplyContext::cacheHit() needs that protection but findPreviouslyCachedEntry() does not. In a sense, collapsed_forwarding_access becomes just one of the many "Can this matching Store entry satisfy our request?" checks that Squid already has to perform.

If a (high-level) caller should not be subject to CF risks but lacks the request structure to test collapsed_forwarding_access, then we may want to add a source code comment to mark that inconsistency.

BTW, addressing this problem may also help us avoid checking collapsed_forwarding_access twice-in-a-row for the same (allowed) request, as the PR code currently does in processExpired() AFAICT.

Reply:
The findPreviouslyCachedEntry() would be protected without any changes:
* In httpMaybeRemovePublic() (as the comment "This protects ready-for-collapsing entries as well") states
* In haveParsedReplyHeaders() it would have no effect since the collapsed entry can't be 'old' and has not got yet a reply.

However, I agree that such relying on occasional circumstances would be unreliable.
I see several different storeGetPublic*() contexts. We need to decide whether the gotten public entry can be used for future collapsing there (and, hence, the new directive could be applicable): there). So we have:
1. For purging(releasing). No.
2. Getting for a transaction. Yes
3. Getting by icp/htcp code (transaction-less entries). No.
4. CF revalidation. Common CF is unrelated here. No.

Also, I looked threw request-less `storeGetPublic()` contexts, I think all of them should not care and check CF-related conditions.